File: lib/json-control/python/serve.py
"""
serve.py — Python HTTP server for the JSON-Control demo site.

Mirror of serve.js: serves the same static files and REST endpoints so
the JS and Python servers are interchangeable (mirror universe).

The control registry is built entirely client-side by build_registry.js —
this server only needs to:
  1. Serve dist/ and src/ as static files with correct MIME types.
  2. Expose GET /api/source/controls/ so build_registry.js can discover files.
  3. Expose GET /api/logs and POST /api/logs for the SQLite log store.

REST API (mirrors serve.js)
───────────────────────────
  GET  /api/source/controls/          List all .js files in src/controls/ (JSON)
  GET  /api/source/controls/:file     Serve src/controls/<file> as text/plain
  GET  /api/source/                   List all .js files in src/ top-level (JSON)
  GET  /api/source/:file              Serve src/<file>.js as text/plain
  GET  /api/logs                      Query log entries (?control= &level= &limit=)
  POST /api/logs                      Insert a log entry { level, context, message, detail? }

Static routes
─────────────
  /              → dist/index.html
  /src/...       → src/... (application/javascript — CRITICAL for native ESM)
  anything else  → dist/<path>

Usage:
  python python/serve.py
  PORT=7777 python python/serve.py
"""

import json
import mimetypes
import os
import sys
import uuid
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

# ── Paths ──────────────────────────────────────────────────────────────────────

_HERE        = Path(__file__).parent
_ROOT        = _HERE.parent
_DIST_DIR    = _ROOT / 'dist'
_SRC_DIR     = _ROOT / 'src'
_CONTROLS_DIR = _SRC_DIR / 'controls'

PORT = int(os.environ.get('PORT', 7777))

# ── MIME types ─────────────────────────────────────────────────────────────────

_MIME = {
    '.html': 'text/html; charset=utf-8',
    '.js':   'application/javascript; charset=utf-8',
    '.mjs':  'application/javascript; charset=utf-8',
    '.json': 'application/json; charset=utf-8',
    '.css':  'text/css; charset=utf-8',
    '.png':  'image/png',
    '.svg':  'image/svg+xml',
    '.ico':  'image/x-icon',
}

_CORS = {
    'Access-Control-Allow-Origin':  '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Cache-Control':                'no-cache',
}

# ── DB import (optional — logs degrade gracefully if unavailable) ──────────────

try:
    sys.path.insert(0, str(_HERE))
    import db as _db
    _db.set_instance_id(str(uuid.uuid4()))
    _DB_AVAILABLE = True
except Exception as _e:
    print(f'[serve.py] db.py not available: {_e}')
    _DB_AVAILABLE = False


# ── Handler ────────────────────────────────────────────────────────────────────

class Handler(BaseHTTPRequestHandler):

    def log_message(self, fmt, *args):
        print(f'  {self.address_string()} {fmt % args}')

    # ── Helpers ───────────────────────────────────────────────────────────────

    def _cors_headers(self, extra=None):
        h = dict(_CORS)
        if extra:
            h.update(extra)
        return h

    def _json(self, status, data):
        body = json.dumps(data).encode()
        self.send_response(status)
        for k, v in self._cors_headers({'Content-Type': 'application/json; charset=utf-8'}).items():
            self.send_header(k, v)
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _serve_file(self, file_path: Path):
        if not file_path.is_file():
            self._json(404, {'error': 'Not found', 'path': str(file_path)})
            return
        ext  = file_path.suffix.lower()
        mime = _MIME.get(ext, 'application/octet-stream')
        data = file_path.read_bytes()
        self.send_response(200)
        for k, v in self._cors_headers({'Content-Type': mime}).items():
            self.send_header(k, v)
        self.send_header('Content-Length', str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _serve_source_text(self, file_path: Path, base_dir: Path):
        """Serve a source file as text/plain (for /api/source/... endpoints)."""
        try:
            resolved = file_path.resolve()
            base_res = base_dir.resolve()
            if not str(resolved).startswith(str(base_res)):
                self._json(403, {'error': 'Forbidden'})
                return
        except Exception:
            self._json(403, {'error': 'Forbidden'})
            return
        if not file_path.is_file():
            self._json(404, {'error': 'Not found'})
            return
        data = file_path.read_bytes()
        self.send_response(200)
        for k, v in self._cors_headers({'Content-Type': 'text/plain; charset=utf-8'}).items():
            self.send_header(k, v)
        self.send_header('Content-Length', str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _read_body(self):
        length = int(self.headers.get('Content-Length', 0))
        return self.rfile.read(length) if length > 0 else b''

    # ── Routing ───────────────────────────────────────────────────────────────

    def do_OPTIONS(self):
        self.send_response(204)
        for k, v in _CORS.items():
            self.send_header(k, v)
        self.end_headers()

    def do_GET(self):
        parsed   = urlparse(self.path)
        url_path = parsed.path
        qs       = parse_qs(parsed.query)

        # ── GET /api/source/controls/ ─────────────────────────────────────────
        if url_path in ('/api/source/controls/', '/api/source/controls'):
            try:
                files = sorted(f for f in os.listdir(_CONTROLS_DIR) if f.endswith('.js'))
                self._json(200, files)
            except Exception as e:
                self._json(500, {'error': str(e)})
            return

        # ── GET /api/source/controls/:file ────────────────────────────────────
        if url_path.startswith('/api/source/controls/'):
            fname = url_path[len('/api/source/controls/'):]
            if '/' in fname or not fname.endswith('.js'):
                self._json(400, {'error': 'Filename must be a single .js file'})
                return
            self._serve_source_text(_CONTROLS_DIR / fname, _CONTROLS_DIR)
            return

        # ── GET /api/source/ ──────────────────────────────────────────────────
        if url_path in ('/api/source/', '/api/source'):
            try:
                files = sorted(
                    f for f in os.listdir(_SRC_DIR)
                    if f.endswith('.js') and (_SRC_DIR / f).is_file()
                )
                self._json(200, files)
            except Exception as e:
                self._json(500, {'error': str(e)})
            return

        # ── GET /api/source/:file ─────────────────────────────────────────────
        if url_path.startswith('/api/source/'):
            fname = url_path[len('/api/source/'):]
            if '/' in fname or not fname.endswith('.js'):
                self._json(400, {'error': 'Filename must be a single .js file with no path separators'})
                return
            self._serve_source_text(_SRC_DIR / fname, _SRC_DIR)
            return

        # ── GET /api/logs ─────────────────────────────────────────────────────
        if url_path == '/api/logs':
            context = (qs.get('control') or qs.get('context') or [None])[0]
            level   = (qs.get('level') or [None])[0]
            limit   = min(int((qs.get('limit') or ['200'])[0]), 1000)
            rows    = _db.query_logs(context, level, limit) if _DB_AVAILABLE else []
            self._json(200, rows)
            return

        # ── GET /src/... — serve ES modules ──────────────────────────────────
        if url_path.startswith('/src/'):
            rel       = url_path[len('/src/'):]
            file_path = (_SRC_DIR / rel).resolve()
            if not str(file_path).startswith(str(_SRC_DIR.resolve())):
                self._json(403, {'error': 'Forbidden'})
                return
            self._serve_file(file_path)
            return

        # ── / or /index.html ──────────────────────────────────────────────────
        if url_path in ('/', '/index.html'):
            self._serve_file(_DIST_DIR / 'index.html')
            return

        # ── Everything else → dist/ ───────────────────────────────────────────
        rel       = url_path.lstrip('/')  or 'index.html'
        file_path = (_DIST_DIR / rel).resolve()
        if not str(file_path).startswith(str(_DIST_DIR.resolve())):
            self._json(403, {'error': 'Forbidden'})
            return
        self._serve_file(file_path)

    def do_POST(self):
        parsed   = urlparse(self.path)
        url_path = parsed.path

        # ── POST /api/logs ────────────────────────────────────────────────────
        if url_path == '/api/logs':
            raw = self._read_body()
            if not raw:
                self._json(400, {'error': 'Empty body'})
                return
            try:
                body = json.loads(raw)
            except Exception:
                self._json(400, {'error': 'Invalid JSON'})
                return

            level   = (body.get('level')   or 'message').lower()
            context = (body.get('context') or 'unknown').strip()
            message = (body.get('message') or '').strip()
            if not message:
                self._json(400, {'error': 'message is required'})
                return

            if level not in _db.VALID_LEVELS if _DB_AVAILABLE else True:
                level = 'message'

            if _DB_AVAILABLE:
                _db.insert_log(
                    level,
                    context,
                    message,
                    detail=body.get('detail'),
                    instance_id=body.get('instance_id'),
                )
                self._json(201, {'ok': True})
            else:
                self._json(503, {'error': 'Logger unavailable (db.py not loaded)'})
            return

        self._json(404, {'error': 'Not found'})


# ── Main ───────────────────────────────────────────────────────────────────────

if __name__ == '__main__':
    server = ThreadingHTTPServer(('0.0.0.0', PORT), Handler)
    print()
    print('JSON-Control Python server running (mirror universe)')
    print()
    print(f'  Index:  http://localhost:{PORT}/')
    print()
    print('  API:')
    print('    GET  /api/source/controls/  list all control files')
    print('    GET  /api/source/controls/:file')
    print('    GET  /api/source/           list top-level src/ files')
    print('    GET  /api/source/:file')
    print('    GET  /api/logs[?control=&level=&limit=]')
    print('    POST /api/logs  { level, context, message, detail? }')
    print()
    print('Press Ctrl-C to stop.')
    print()
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\nStopped.')

File: lib/json-control/serve.js
#!/usr/bin/env node
/**
 * serve.js — static dev server for the JSON-Control demo site.
 *
 * Serves both dist/ (the HTML shell) and src/ (the ES module source files).
 * The browser loads controls natively via <script type="module"> — no bundler.
 *
 * On startup, buildControlRegistry() imports every control module, reads each
 * .meta static property, and reads the source file from disk. The resulting
 * object is keyed by control_type and exposed as GET /api/control_registry.
 *
 * NOTE: shell.js now builds the registry client-side via build_registry.js and
 * /api/source/controls/. The /api/control_registry endpoint is retained as a
 * server-side convenience and for parity with direct API consumers.
 *
 * REST API
 * ────────
 *   GET  /api/control_registry              Full control registry (JSON)
 *                                             control_registry[controlType] = {
 *                                               controlType, category, description,
 *                                               defaultConfig, tests, srcCode }
 *   GET  /api/source/controls/              List all .js files in src/controls/
 *   GET  /api/source/controls/:file         Source of src/controls/<file> (text/plain)
 *   GET  /api/source/                       List all .js files in src/ top-level
 *   GET  /api/source/:file                  Source of any .js in src/ top-level (text/plain)
 *   GET  /api/logs                          Query log entries from SQLite
 *                                             ?control=  filter by control name (= context)
 *                                             ?context=  same as ?control
 *                                             ?level=    DEBUG|MESSAGE|WARNING|CRITICAL
 *                                             ?limit=    max rows (default 200, cap 1000)
 *   POST /api/logs                          Append a log entry to SQLite
 *                                             body JSON: { level, context, message, detail? }
 *
 * Static routes
 * ─────────────
 *   /            → dist/index.html
 *   /src/...     → src/...   (ES modules, served with application/javascript)
 *   anything else → dist/<path>
 *
 * Usage:
 *   node serve.js
 *   PORT=8080 node serve.js
 */

import http              from 'http'
import fs                from 'fs'
import path              from 'path'
import { randomUUID }    from 'crypto'
import { createRequire } from 'module'
import { fileURLToPath } from 'url'

const INSTANCE_ID = randomUUID()

// ── Per-startup build version ─────────────────────────────────────────────────
// A fresh token is minted every time the server process starts.  JS modules
// are served under /src/v{BUILD_VER}/... so every deploy gets unique URLs,
// bypassing any intermediate proxy cache (including the Replit preview proxy)
// that might ignore Cache-Control: no-store for native ESM modules.
const BUILD_VER = Date.now().toString(36)

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const require   = createRequire(import.meta.url)

const rootDir     = __dirname
const distDir     = path.join(rootDir, 'dist')
const srcDir      = path.join(rootDir, 'src')
const controlsDir = path.join(srcDir, 'controls')
const DB_PATH     = path.join(rootDir, 'test-results.db')

// ── Control registry (built on startup) ──────────────────────────────────────

var control_registry = {}

async function buildControlRegistry () {
  var files = fs.readdirSync(controlsDir)
    .filter(function (f) { return f.endsWith('.js') && f !== 'index.js' })
    .sort()

  for (var file of files) {
    try {
      var srcCode = fs.readFileSync(path.join(controlsDir, file), 'utf8')
      var mod     = await import(path.join(controlsDir, file))
      var Ctrl    = Object.values(mod).find(function (v) {
        return typeof v === 'function' && v.meta
      })
      if (!Ctrl) {
        console.warn('control_registry: no .meta found in', file)
        continue
      }
      var meta  = Ctrl.meta
      var tests = (meta.tests || []).map(function (t) {
        return { name: t.name, fn: t.fn.toString() }
      })
      control_registry[meta.controlType] = {
        controlType:   meta.controlType,
        category:      meta.category,
        description:   meta.description,
        defaultConfig: meta.defaultConfig,
        tests:         tests,
        srcCode:       srcCode,
      }
    } catch (err) {
      console.error('control_registry: failed to load ' + file + ':', err.message)
    }
  }
}

// ── Logger (CJS — used for writes) ───────────────────────────────────────────
var _logger = null
function getLogger () {
  if (_logger) return _logger
  try { _logger = require('./db/logger.cjs') } catch (_) {}
  return _logger
}

// ── SQLite read connection ────────────────────────────────────────────────────
var _db = null
function getDb () {
  if (_db) return _db
  try {
    var Database = require('better-sqlite3')
    if (!fs.existsSync(DB_PATH)) return null
    _db = new Database(DB_PATH, { readonly: true })
    return _db
  } catch (_) { return null }
}

function queryLogs (context, level, limit) {
  var db = getDb()
  if (!db) return []
  try {
    var sql  = 'SELECT id, logged_at, level, context, message, detail FROM logs WHERE 1=1'
    var args = []
    if (context) { sql += ' AND context = ?'; args.push(context) }
    if (level)   { sql += ' AND level = ?';   args.push(level.toUpperCase()) }
    sql += ' ORDER BY logged_at DESC LIMIT ?'
    args.push(limit)
    var rows = db.prepare(sql).all(...args)
    return rows.map(function (r) {
      var detail = r.detail
      if (detail) { try { detail = JSON.parse(detail) } catch (_) {} }
      return { id: r.id, logged_at: r.logged_at, level: r.level, context: r.context, message: r.message, detail }
    })
  } catch (_) { return [] }
}

// ── Helpers ───────────────────────────────────────────────────────────────────
const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js':   'application/javascript; charset=utf-8',
  '.mjs':  'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.css':  'text/css; charset=utf-8',
  '.png':  'image/png',
  '.svg':  'image/svg+xml',
  '.ico':  'image/x-icon',
}

function json (res, status, data) {
  var body = JSON.stringify(data)
  res.writeHead(status, {
    'Content-Type':  'application/json; charset=utf-8',
    'Cache-Control': 'no-cache',
    'Access-Control-Allow-Origin': '*',
  })
  res.end(body)
}

function serveSource (res, filePath, baseDir) {
  if (!filePath.startsWith(baseDir)) {
    return json(res, 403, { error: 'Forbidden' })
  }
  fs.stat(filePath, function (err, stat) {
    if (err || !stat.isFile()) {
      return json(res, 404, { error: 'Not found', path: filePath.slice(rootDir.length) })
    }
    res.writeHead(200, {
      'Content-Type':  'text/plain; charset=utf-8',
      'Cache-Control': 'no-cache',
      'Access-Control-Allow-Origin': '*',
    })
    fs.createReadStream(filePath).pipe(res)
  })
}

function serveFile (res, filePath) {
  fs.stat(filePath, function (statErr, stat) {
    if (statErr || !stat.isFile()) {
      res.writeHead(404, { 'Content-Type': 'text/plain' })
      res.end('Not found: ' + filePath)
      return
    }
    var ext  = path.extname(filePath).toLowerCase()
    var mime = MIME[ext] || 'application/octet-stream'
    res.writeHead(200, {
      'Content-Type':  mime,
      'Cache-Control': 'no-store, no-cache, must-revalidate',
      'Pragma':        'no-cache',
      'Access-Control-Allow-Origin': '*',
    })
    fs.createReadStream(filePath).pipe(res)
  })
}

function readBody (req, cb) {
  var chunks = []
  req.on('data', function (c) { chunks.push(c) })
  req.on('end', function () { cb(Buffer.concat(chunks).toString('utf8')) })
  req.on('error', function () { cb(null) })
}

// ── Server ────────────────────────────────────────────────────────────────────
const PORT = parseInt(process.env.PORT || '7777', 10)

const server = http.createServer(function (req, res) {
  var urlPath = req.url.split('?')[0]
  var method  = req.method.toUpperCase()

  // OPTIONS (CORS preflight)
  if (method === 'OPTIONS') {
    res.writeHead(204, {
      'Access-Control-Allow-Origin':  '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type',
    })
    res.end()
    return
  }

  // ── GET /api/control_registry ─────────────────────────────────────────────
  if (method === 'GET' && urlPath === '/api/control_registry') {
    return json(res, 200, control_registry)
  }

  // ── GET /api/source/controls/          → list all .js files in src/controls/
  // ── GET /api/source/controls/:file     → source of one file
  if (method === 'GET' && urlPath.startsWith('/api/source/controls')) {
    var file = urlPath.slice('/api/source/controls'.length).replace(/^\//, '')

    if (!file) {
      try {
        var entries = fs.readdirSync(controlsDir).filter(function (f) { return f.endsWith('.js') }).sort()
        return json(res, 200, entries)
      } catch (e) {
        return json(res, 500, { error: e.message })
      }
    }

    if (file.includes('/') || !file.endsWith('.js')) {
      return json(res, 400, { error: 'Filename must be a single .js file' })
    }
    serveSource(res, path.join(controlsDir, file), controlsDir)
    return
  }

  // ── GET /api/source/          → list all .js files in src/ top-level
  // ── GET /api/source/:file     → source of one file
  if (method === 'GET' && urlPath.startsWith('/api/source')) {
    var file2 = urlPath.slice('/api/source'.length).replace(/^\//, '')

    if (!file2) {
      try {
        var entries2 = fs.readdirSync(srcDir).filter(function (f) { return f.endsWith('.js') && !fs.statSync(path.join(srcDir, f)).isDirectory() }).sort()
        return json(res, 200, entries2)
      } catch (e) {
        return json(res, 500, { error: e.message })
      }
    }

    if (file2.includes('/') || !file2.endsWith('.js')) {
      return json(res, 400, { error: 'Filename must be a single .js file with no path separators' })
    }
    serveSource(res, path.join(srcDir, file2), srcDir)
    return
  }

  // ── /api/logs ─────────────────────────────────────────────────────────────
  if (urlPath === '/api/logs') {

    // GET — query log entries
    if (method === 'GET') {
      var params  = new URLSearchParams(req.url.split('?')[1] || '')
      var context = params.get('control') || params.get('context') || null
      var level   = params.get('level')   || null
      var limit   = Math.min(parseInt(params.get('limit') || '200', 10), 1000)
      return json(res, 200, queryLogs(context, level, limit))
    }

    // POST — insert a log entry
    if (method === 'POST') {
      readBody(req, function (raw) {
        if (!raw) return json(res, 400, { error: 'Empty body' })
        var body
        try { body = JSON.parse(raw) } catch (_) {
          return json(res, 400, { error: 'Invalid JSON' })
        }
        var level   = (body.level   || 'message').toLowerCase()
        var context = (body.context || 'unknown').trim()
        var message = (body.message || '').trim()
        if (!message) return json(res, 400, { error: 'message is required' })

        var validLevels = ['debug', 'message', 'warning', 'critical']
        if (!validLevels.includes(level)) level = 'message'

        var lg = getLogger()
        if (lg) {
          var logOpts = body.instance_id ? { instanceId: body.instance_id } : undefined
          lg.log(level, context, message, body.detail !== undefined ? body.detail : undefined, logOpts)
          return json(res, 201, { ok: true })
        } else {
          return json(res, 503, { error: 'Logger unavailable (SQLite not loaded)' })
        }
      })
      return
    }

    return json(res, 405, { error: 'Method not allowed' })
  }

  // ── /src/v{BUILD_VER}/... — versioned ES module source files ────────────────
  // Every server restart produces a new BUILD_VER.  The HTML entry point injects
  // this version into the <script src="..."> URL so all modules in the graph are
  // served under a unique path, bypassing proxy and browser module-map caching.
  var SRC_VER_PREFIX = '/src/v' + BUILD_VER + '/'
  if (urlPath.startsWith(SRC_VER_PREFIX)) {
    var rel   = urlPath.slice(SRC_VER_PREFIX.length)
    var file3 = path.join(srcDir, rel)
    if (!file3.startsWith(srcDir)) { res.writeHead(403); res.end('Forbidden'); return }
    serveFile(res, file3)
    return
  }

  // ── /src/... — legacy unversioned route (still works for direct access) ───
  if (urlPath.startsWith('/src/')) {
    var relLeg   = urlPath.slice('/src/'.length)
    var file3Leg = path.join(srcDir, relLeg)
    if (!file3Leg.startsWith(srcDir)) { res.writeHead(403); res.end('Forbidden'); return }
    serveFile(res, file3Leg)
    return
  }

  // ── / or /index.html → dist/index.html (with versioned script src) ────────
  // The shell.js script tag is rewritten to point at the versioned path so that
  // all relative imports resolve under the same /src/v{ver}/ prefix, giving
  // every module a fresh URL on each server restart.
  if (urlPath === '/' || urlPath === '/index.html') {
    var htmlPath = path.join(distDir, 'index.html')
    try {
      var htmlSrc = fs.readFileSync(htmlPath, 'utf8')
      var htmlOut = htmlSrc.replace(
        /src="[^"]*shell\.js"/,
        'src="/src/v' + BUILD_VER + '/shell.js"'
      )
      res.writeHead(200, {
        'Content-Type':  'text/html; charset=utf-8',
        'Cache-Control': 'no-store, no-cache, must-revalidate',
        'Pragma':        'no-cache',
        'Access-Control-Allow-Origin': '*',
      })
      res.end(htmlOut)
    } catch (e) {
      res.writeHead(500, { 'Content-Type': 'text/plain' })
      res.end('Failed to serve index.html: ' + e.message)
    }
    return
  }

  // ── Everything else → dist/ ───────────────────────────────────────────────
  var rel4  = urlPath.replace(/^\/+/, '') || 'index.html'
  var file4 = path.join(distDir, rel4)
  if (!file4.startsWith(distDir)) { res.writeHead(403); res.end('Forbidden'); return }
  serveFile(res, file4)
})

// ── Startup ───────────────────────────────────────────────────────────────────
;(async function main () {
  var lg = getLogger()
  if (lg) lg.setInstanceId(INSTANCE_ID)

  await buildControlRegistry()
  var count = Object.keys(control_registry).length
  server.listen(PORT, '0.0.0.0', function () {
    console.log()
    console.log('JSON-Control demo server running (ES modules, no bundler)')
    console.log('Control registry: ' + count + ' controls loaded')
    console.log('Build version:    ' + BUILD_VER + '  (JS served at /src/v' + BUILD_VER + '/)')
    console.log()
    console.log('  Index:  http://localhost:' + PORT + '/')
    console.log()
    console.log('  API:')
    console.log('    GET  /api/control_registry')
    console.log('    GET  /api/source/           list top-level src/ files')
    console.log('    GET  /api/source/:file      source of src/<file>.js')
    console.log('    GET  /api/source/controls/  list all control files')
    console.log('    GET  /api/source/controls/:file')
    console.log('    GET  /api/logs[?control=&level=&limit=]')
    console.log('    POST /api/logs  { level, context, message, detail? }')
    console.log()
    console.log('Press Ctrl-C to stop.')
  })
})()
